device/futhark/src/cca.cpp

```diff
@@ -0,0 +1,149 @@
+        futhark_new_i64_1d(ctx, host_channel0, total_cells);
+    struct futhark_i64_1d* in_channel1 =
+        futhark_new_i64_1d(ctx, host_channel1, total_cells);
+    struct futhark_f32_1d* in_activation =
+        futhark_new_f32_1d(ctx, host_activation, total_cells);
+
+    delete[] host_event;
+    delete[] host_geometry;
+    delete[] host_channel0;
+    delete[] host_channel1;
+    delete[] host_activation;
+
+    struct futhark_u64_1d* out_event;
+    struct futhark_u64_1d* out_geometry;
+    struct futhark_f32_1d* out_position0;
+    struct futhark_f32_1d* out_position1;
+    struct futhark_f32_1d* out_variance0;
+    struct futhark_f32_1d* out_variance1;
+
+    int r = futhark_entry_cells_to_measurements_entry(
+        ctx, &out_event, &out_geometry, &out_position0, &out_position1,
+        &out_variance0, &out_variance1, in_event, in_geometry, in_channel0,
+        in_channel1, in_activation);
+
+    if (r == FUTHARK_PROGRAM_ERROR) {
+        throw std::runtime_error(
+            "Futhark program exited due to a programming error.");
+    } else if (r == FUTHARK_OUT_OF_MEMORY) {
+        throw std::runtime_error(
+            "Futhark program exited due to lack of allocatable memory.");
+    } else if (r != FUTHARK_SUCCESS) {
+        std::stringstream ss;
+        ss << "Futhark program exited with unknown non-zero return code " << r
+           << ".";
+
+        throw std::runtime_error(ss.str());
+    }
+
+    futhark_free_u64_1d(ctx, in_event);
+    futhark_free_u64_1d(ctx, in_geometry);
+    futhark_free_i64_1d(ctx, in_channel0);
+    futhark_free_i64_1d(ctx, in_channel1);
+    futhark_free_f32_1d(ctx, in_activation);
+
+    int64_t total_measurements = *futhark_shape_u64_1d(ctx, out_event);
+
+    uint64_t* host_out_event = new uint64_t[total_measurements];
+    uint64_t* host_out_geometry = new uint64_t[total_measurements];
+    float* host_out_position0 = new float[total_measurements];
+    float* host_out_position1 = new float[total_measurements];
+    float* host_out_variance0 = new float[total_measurements];
+    float* host_out_variance1 = new float[total_measurements];
+
+    futhark_values_u64_1d(ctx, out_event, host_out_event);
+    futhark_values_u64_1d(ctx, out_geometry, host_out_geometry);
+    futhark_values_f32_1d(ctx, out_position0, host_out_position0);
+    futhark_values_f32_1d(ctx, out_position1, host_out_position1);
+    futhark_values_f32_1d(ctx, out_variance0, host_out_variance0);
+    futhark_values_f32_1d(ctx, out_variance1, host_out_variance1);
+
+    futhark_free_u64_1d(ctx, out_event);
+    futhark_free_u64_1d(ctx, out_geometry);
+    futhark_free_f32_1d(ctx, out_position0);
+    futhark_free_f32_1d(ctx, out_position1);
+    futhark_free_f32_1d(ctx, out_variance0);
+    futhark_free_f32_1d(ctx, out_variance1);
+
+    output_type out;
+    vecmem::host_memory_resource mem;
+
+    for (std::size_t i = 0; i < data.size(); ++i) {
+        vecmem::vector<measurement> v(&mem);
```

Comment:
You should reserve an appropriate size for the vector. I think over-reserving is still a bit better performance-wise than just letting `std::vector` reserve the memory in multiple iterations.